build: opentelemetry 0.32 broke SpanBuilder usage in `trace-ot` feature

## Summary

Building with the `trace-ot` feature fails on `main` after the workspace bump
to `opentelemetry = "0.32"` (and `opentelemetry_sdk = "0.32"`,
`tracing-opentelemetry = "0.33"`). The error is masked because the default
CI matrix doesn't enable `trace-ot`, but any developer running
`cargo clippy -p freenet --all-features -- -D warnings` (or building with
the feature explicitly) hits:

```
error[E0560]: struct `opentelemetry::trace::SpanBuilder` has no field named `span_id`
    --> crates/core/src/tracing.rs:2223:17
     |
2223 |                 span_id: Some(trace::SpanId::from_bytes(span_id)),

error[E0560]: struct `opentelemetry::trace::SpanBuilder` has no field named `trace_id`
    --> crates/core/src/tracing.rs:2224:17
```

## Root cause

`opentelemetry 0.32`'s `SpanBuilder` removed the public `span_id` and
`trace_id` fields. The exported fields are now only `span_kind`, `name`,
`start_time`, `attributes`, `events`, `links`. Span and trace IDs are no
longer overridable on the builder — the SDK's `id_generator.new_span_id()`
always issues a fresh span_id, and the `trace_id` is inherited from a
parent `SpanContext` placed in the `Context`, otherwise generated.

This is a meaningful semantic change for `OTSpan::new` in
`crates/core/src/tracing.rs` (around line 2212), which was deriving both
IDs deterministically from the transaction bytes so multiple spans across
a transaction would share `trace_id`/`span_id`.

## Proposed fix shape

Build a remote parent `SpanContext` carrying the transaction-derived
`trace_id` (and the transaction-derived `span_id` as the parent span_id),
then call `tracer.build_with_context(builder, &parent_cx)` so the new
span inherits the trace_id. The new span itself receives a
generator-assigned span_id (this is a wire-format change in the exported
spans — the per-span ID is no longer transaction-derived).

```rust
use opentelemetry::{
    Context, KeyValue, global,
    trace::{self, Span, TraceContextExt, TraceFlags, TraceState},
};

let parent_span_ctx = trace::SpanContext::new(
    trace::TraceId::from_bytes(tx_bytes),
    trace::SpanId::from_bytes(span_id_bytes),
    TraceFlags::SAMPLED,
    true,
    TraceState::default(),
);
let parent_cx = Context::new().with_remote_span_context(parent_span_ctx);
let builder = trace::SpanBuilder::from_name(transaction.transaction_type().description())
    .with_start_time(start_time)
    .with_attributes(vec![
        KeyValue::new("transaction", transaction.to_string()),
        KeyValue::new("tx_type", transaction.transaction_type().description()),
    ]);
let inner = tracer.build_with_context(builder, &parent_cx);
```

I verified this compiles cleanly against the workspace's pinned
`opentelemetry 0.32.0` locally on a worktree, but did not land it because
the change is out-of-scope for the in-progress routing PR I was working
on.

## Adjacent issue (same file, surfaced by the same clippy run)

While `--all-features -- -D warnings`, two additional wildcard-match
errors appear on the `trace-ot`-gated paths in the same file:

- `crates/core/src/tracing.rs:1559` — `_ => false,` inside
  `NetLogMessage::span_completed`
- `crates/core/src/tracing.rs:1626` — `_ => None,` inside the
  `impl From<&NetLogMessage> for Option<Vec<opentelemetry::KeyValue>>`

These violate `clippy::wildcard_enum_match_arm` because new `EventKind`
variants have landed since 2023. Either expand the matches or add
`#[allow(clippy::wildcard_enum_match_arm)]` at the enclosing `fn` level
with a comment (the existing `tracing.rs` site at lines 2899/2911 uses
the latter pattern).

## Why this matters

- Anyone enabling `trace-ot` (developers iterating on OTLP/Jaeger
  exporters) gets a hard build break.
- A `--all-features` clippy run — which is a common pre-PR sanity check —
  fails on `main`, which makes it harder to distinguish "my PR broke
  things" from "main was already broken".

## Reproduction

```bash
cargo clippy -p freenet --all-features -- -D warnings
```

## Suggested action

Fix `OTSpan::new` per the snippet above, add the `#[allow]` on the two
wildcard sites, run `cargo test -p freenet --features trace-ot --lib tracing::`
to confirm the tracing tests still pass, then add a CI job (or extend an
existing matrix) that builds with `trace-ot` enabled so this can't drift
silently again.

[AI-assisted - Claude]

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -401,6 +401,16 @@ jobs:
           CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUSTFLAGS: -C link-arg=-fuse-ld=mold
         run: cargo clippy --locked -- -D warnings
 
+      # Guards the opt-in `trace-ot` OpenTelemetry feature against silent
+      # rot: it is excluded from the default build, so a refactor can break
+      # its #[cfg(feature = "trace-ot")] code paths without any other CI job
+      # noticing. See #4225.
+      - name: clippy (trace-ot feature)
+        env:
+          CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER: clang
+          CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUSTFLAGS: -C link-arg=-fuse-ld=mold
+        run: cargo clippy --locked -p freenet --features trace-ot -- -D warnings
+
   # Windows compile check - catches missing imports, feature flags, and
   # platform-specific errors that Linux CI misses. Added after #3685 where
   # PRs #3669/#3680 introduced Windows-only code that failed to compile,

--- a/crates/core/src/tracing/metrics_client.rs
+++ b/crates/core/src/tracing/metrics_client.rs
@@ -644,5 +644,9 @@ mod opentelemetry_tracer {
     }
 }
 
+// `pub(crate)` (not `pub(super)`): `OTEventRegister` is re-exported again as
+// `pub(crate)` from `tracing.rs` and constructed in `node.rs` / `testing_impl`,
+// so it must be visible crate-wide. The narrower `pub(super)` broke the
+// `trace-ot` build after the tracing module split (see #4225).
 #[cfg(feature = "trace-ot")]
-pub(super) use opentelemetry_tracer::OTEventRegister;
+pub(crate) use opentelemetry_tracer::OTEventRegister;

--- a/crates/core/src/tracing/register.rs
+++ b/crates/core/src/tracing/register.rs
@@ -1553,8 +1553,14 @@ impl NetLogMessage {
     /// Signals whether this message closes a transaction span.
     ///
     /// In case of isolated events where the span is not being tracked it should return true.
+    // Terminal variants above complete a span; the wildcard is deliberate so
+    // non-span events (and any future variant) are treated as not-completing.
+    // `pub(crate)`: called from the sibling `metrics_client` module's
+    // OpenTelemetry span processing. Was private, which broke the `trace-ot`
+    // build after the tracing module split (see #4225).
     #[cfg(feature = "trace-ot")]
-    fn span_completed(&self) -> bool {
+    #[allow(clippy::wildcard_enum_match_arm)]
+    pub(crate) fn span_completed(&self) -> bool {
         match &self.kind {
             EventKind::Connect(ConnectEvent::Finished { .. }) => true,
             EventKind::Connect(_) => false,
@@ -1568,6 +1574,10 @@ impl NetLogMessage {
                 | SubscribeEvent::SubscribeTimeout { .. },
             ) => true,
             EventKind::Subscribe(_) => false,
+            EventKind::Update(
+                UpdateEvent::UpdateSuccess { .. } | UpdateEvent::UpdateFailure { .. },
+            ) => true,
+            EventKind::Update(_) => false,
             _ => false,
         }
     }
@@ -1586,6 +1596,9 @@ impl<'a> From<NetEventLog<'a>> for NetLogMessage {
 
 #[cfg(feature = "trace-ot")]
 impl<'a> From<&'a NetLogMessage> for Option<Vec<opentelemetry::KeyValue>> {
+    // Wildcard is deliberate: only the event kinds mapped above carry
+    // span attributes; every other event (and any new variant) maps to None.
+    #[allow(clippy::wildcard_enum_match_arm)]
     fn from(msg: &'a NetLogMessage) -> Self {
         use opentelemetry::KeyValue;
         let map: Option<Vec<KeyValue>> = match &msg.kind {
@@ -2054,5 +2067,81 @@ mod eventlog_backpressure_tests {
         )
         .await
         .expect("register_events must return promptly on a closed channel");
+    }
+}
+
+#[cfg(all(test, feature = "trace-ot"))]
+mod span_completed_tests {
+    use super::*;
+
+    fn msg(kind: EventKind) -> NetLogMessage {
+        NetLogMessage {
+            datetime: Utc::now(),
+            tx: *Transaction::NULL,
+            peer_id: PeerId::random(),
+            kind,
+        }
+    }
+
+    fn contract_key() -> ContractKey {
+        ContractKey::from_id_and_code(ContractInstanceId::new([1u8; 32]), CodeHash::new([2u8; 32]))
+    }
+
+    // Regression for #4225: UpdateSuccess / UpdateFailure are terminal UPDATE
+    // events that close their OpenTelemetry span. They previously fell through
+    // the `_ => false` arm of `span_completed`, so their spans were never
+    // marked complete and leaked under the `trace-ot` feature.
+    #[test]
+    fn update_terminal_events_complete_span() {
+        let tx = Transaction::new::<crate::operations::update::UpdateMsg>();
+        let key = contract_key();
+
+        let success = msg(EventKind::Update(UpdateEvent::UpdateSuccess {
+            id: tx,
+            requester: PeerKeyLocation::random(),
+            target: PeerKeyLocation::random(),
+            key,
+            timestamp: 0,
+            state_hash_before: None,
+            state_hash_after: None,
+            state_size: None,
+        }));
+        assert!(
+            success.span_completed(),
+            "UpdateSuccess must close its span"
+        );
+
+        let failure = msg(EventKind::Update(UpdateEvent::UpdateFailure {
+            id: tx,
+            requester: PeerKeyLocation::random(),
+            target: PeerKeyLocation::random(),
+            key,
+            reason: OperationFailure::Timeout,
+            elapsed_ms: 0,
+            timestamp: 0,
+        }));
+        assert!(
+            failure.span_completed(),
+            "UpdateFailure must close its span"
+        );
+    }
+
+    // A non-terminal UPDATE event must NOT close the span — guards against
+    // over-broadly classifying the whole `Update` group as span-completing.
+    #[test]
+    fn non_terminal_update_event_does_not_complete_span() {
+        let tx = Transaction::new::<crate::operations::update::UpdateMsg>();
+
+        let request = msg(EventKind::Update(UpdateEvent::Request {
+            id: tx,
+            requester: PeerKeyLocation::random(),
+            key: contract_key(),
+            target: PeerKeyLocation::random(),
+            timestamp: 0,
+        }));
+        assert!(
+            !request.span_completed(),
+            "UpdateEvent::Request is not terminal and must not close the span"
+        );
     }
 }

